src/igraph/drawing/matplotlib/edge.py

```diff
@@ -55,216 +61,343 @@ class VisualEdgeBuilder(AttributeCollectorBase):
 
         return VisualEdgeBuilder
 
+    def build_patch(self, edge, src_vertex, dest_vertex):
+        art = mpl.patches.PathPatch(
+            mpl.path.Path([[0, 0]]),
+            edgecolor=edge.color,
+            facecolor=edge.color if src_vertex != dest_vertex else "none",
+            linewidth=edge.width,
+            zorder=edge.zorder,
+            clip_on=True,
+        )
+        return art
+
+    # The following two methods are replaced by dummy functions, the rest is
+    # taken care of in EdgeCollection for efficiency
     def draw_directed_edge(self, edge, src_vertex, dest_vertex):
-        if src_vertex == dest_vertex:
-            return self.draw_loop_edge(edge, src_vertex)
+        return self.build_patch(edge)
 
-        ax = self.context
-        (x1, y1), (x2, y2) = src_vertex.position, dest_vertex.position
-        (x_src, y_src), (x_dest, y_dest) = src_vertex.position, dest_vertex.position
+    def draw_undirected_edge(self, edge, src_vertex, dest_vertex):
+        return self.build_patch(edge)
+
+
+class EdgeCollection(PatchCollection):
+    def __init__(self, *args, **kwargs):
+        kwargs["match_original"] = True
+        self._visual_vertices = kwargs.pop("visual_vertices", None)
```

Comment:
Here and equivalent in the other collections, there is another way to do this, i.e. passing the parent `GraphArtist` straight as an attribute of this class, and grab individual properties when needed